storage/cache/subtree_cache.go

```diff
@@ -436,3 +455,39 @@ func PopulateLogSubtreeNodes(treeHasher merkle.TreeHasher) storage.PopulateSubtr
 		return nil
 	}
 }
+
+// PrepareMapSubtreeWrite prepares a map subtree for writing. For maps the internal
+// nodes are never written to storage and are thus always cleared
+func PrepareMapSubtreeWrite() storage.PrepareSubtreeWriteFunc {
+	return func(st *storagepb.SubtreeProto) error {
+		st.InternalNodes = nil
+		return nil
+	}
+}
+
+// PrepareLogSubtreeWrite prepares a log subtree for writing. If the subtree is fully
+// populated the internal nodes are cleared. Otherwise they are written.
+//
+// To see why this is necessary consider the case where a tree has a single full subtree
+// and then an additional leaf is added.
+//
+// This causes an extra level to be added to the tree with an internal node that is a hash
+// of the root of the left full subtree and the new leaf. Thus the internal nodes cannot be
```

Comment:
"(note that the leaves are stored at level 0, which will be in a lower stratum in this example)"